File: apps/go/internal/sqid/sqid_test.go
package sqid

import (
	"errors"
	"math"
	"strings"
	"testing"
)

const testAlphabet = "k3G7QAe51FfL2rl4wRxyOzZbnucItJ8hgSpEmvNiHqKMWXdVaCDBjT0YoU6P9s"

func newCodec(t *testing.T) *Codec {
	t.Helper()

	c, err := New(testAlphabet)
	if err != nil {
		t.Fatal(err)
	}

	return c
}

func TestRoundTrip(t *testing.T) {
	c := newCodec(t)

	for _, id := range []int64{0, 1, 2, 9, 10, 99, 100, 1000, 123456, math.MaxInt32, math.MaxInt64} {
		encoded, err := c.Encode(id)
		if err != nil {
			t.Fatalf("Encode(%d): %v", id, err)
		}

		if len(encoded) < minLength {
			t.Errorf("Encode(%d) = %q, shorter than %d", id, encoded, minLength)
		}

		decoded, err := c.Decode(encoded)
		if err != nil {
			t.Fatalf("Decode(%q): %v", encoded, err)
		}

		if decoded != id {
			t.Errorf("Decode(Encode(%d)) = %d", id, decoded)
		}
	}
}

func TestDecodeRejects(t *testing.T) {
	c := newCodec(t)

	valid, err := c.Encode(42)
	if err != nil {
		t.Fatal(err)
	}

	cases := map[string]string{
		"empty":            "",
		"zero":             "0",
		"out of alphabet":  "!!!!!!",
		"padded":           valid + " ",
		"absurdly long":    strings.Repeat(testAlphabet, 100),
		"trailing garbage": valid + "zzzz",
	}

	for name, input := range cases {
		if _, err := c.Decode(input); !errors.Is(err, ErrInvalid) {
			t.Errorf("Decode(%s = %q) did not reject", name, input)
		}
	}
}

// Decode is lenient about non-canonical forms, so anything it accepts must re-encode to itself
func TestDecodeRejectsNonCanonical(t *testing.T) {
	c := newCodec(t)

	valid, err := c.Encode(7)
	if err != nil {
		t.Fatal(err)
	}

	for _, padding := range []string{testAlphabet[:1], testAlphabet[:2], testAlphabet[:3]} {
		input := padding + valid

		if input == valid {
			continue
		}

		id, err := c.Decode(input)
		if err == nil {
			t.Errorf("Decode(%q) = %d, accepted a non-canonical form of %q", input, id, valid)
		}
	}
}

func TestEncodeRejectsNegative(t *testing.T) {
	c := newCodec(t)

	if _, err := c.Encode(-1); !errors.Is(err, ErrInvalid) {
		t.Error("Encode(-1) did not reject")
	}
}

func TestNewRejectsBadAlphabet(t *testing.T) {
	for name, alphabet := range map[string]string{
		"empty":     "",
		"too short": "abc",
		"repeated":  strings.Repeat("a", 60),
	} {
		if _, err := New(alphabet); err == nil {
			t.Errorf("New(%s) accepted", name)
		}
	}
}

File: apps/js/src/lib/sqid.test.ts
import { expect, test, vi } from "vite-plus/test";

import { decode, encode } from "#/lib/sqid";

const ALPHABET =
	"k3G7QAe51FfL2rl4wRxyOzZbnucItJ8hgSpEmvNiHqKMWXdVaCDBjT0YoU6P9s";
const MIN_LENGTH = 6;
const TOO_SHORT = "sqid alphabet must be at least 16 characters";
const NOT_AN_ID = "sqid ids must be non-negative safe integers";

process.env["SQIDS_ALPHABET"] = ALPHABET;

test("round-trips across a wide id range", () => {
	const ids = [0, 1, 2, 9, 10, 99, 100, 1000, 123456, 2147483647];

	expect(ids.map((id) => decode(encode(id)))).toStrictEqual(ids);
	expect(ids.filter((id) => encode(id).length < MIN_LENGTH)).toStrictEqual([]);
});

test("rejects garbage and non-canonical forms", () => {
	const valid = encode(42);

	const cases: Record<string, string> = {
		"empty": "",
		"zero": "0",
		"out of alphabet": "!!!!!!",
		"padded": `${valid} `,
		"absurdly long": ALPHABET.repeat(100),
		"trailing garbage": `${valid}zzzz`,
		"non-canonical": ALPHABET.slice(0, 1) + valid,
	};

	const accepted = Object.entries(cases)
		.filter(([, input]) => decode(input) !== undefined)
		.map(([name]) => name);

	expect(accepted).toStrictEqual([]);
});

test("rejects a negative id", () => {
	expect(() => encode(-1)).toThrow(NOT_AN_ID);
});

test("rejects an alphabet short enough to fall back to the public default", async () => {
	const before = process.env["SQIDS_ALPHABET"];

	try {
		for (const alphabet of ["", "abc"]) {
			vi.resetModules();
			process.env["SQIDS_ALPHABET"] = alphabet;

			const fresh = await import("#/lib/sqid");

			expect(() => fresh.encode(1)).toThrow(TOO_SHORT);
		}
	} finally {
		process.env["SQIDS_ALPHABET"] = before;
		vi.resetModules();
	}
});
